### ISSUE TITLE
Possible improvment to the benchmark

### ISSUE THREAD

**Issue by X-Ryl669** [project-x51/esp32-s3-memorycopy#2]
While reading the code, I'm spotting 3 potentials issues:

1. Your code should be placed in IRAM (using the `IRAM_ATTR` [macro](https://docs.espressif.com/projects/esp-idf/en/stable/esp32s2/api-guides/memory-types.html#how-to-place-code-in-iram)) so it's not evicted from cache when the PSRAM is loading (and the time code doesn't measure slowness of SPI Flash reading).
2. I'm not sure you're testing what you expect to test. The IRAM->IRAM test isn't doing what you think, since you've allocated the buffer with MALLOC_CAP_DMA and not MALLOC_CAP_32BIT. It's only with the latter that you *may* get an IRAM buffer (see [here](https://docs.espressif.com/projects/esp-idf/en/stable/esp32s2/api-reference/system/mem_alloc.html#dma-capable-memory)). The other flag won't give you IRAM (they might give you DRAM, so you're measuring the SRAM speed here). Please notice that the memory bus of ESP32-S3 is 240MHz (32 bits, so a maximum theorical bandwidth is 960MB/s, a copy should halve that bandwidth so at best, you *should* get 480MB/s). I don't understand how you could get more than that from the results you've shown. The ESP32 doesn't do any DDR like behavior on the SRAM, only the PSRAM supports it. 
If your results are confirmed, then this means that the DSP/PIE can access the SRAM in larger format than 32 bits (and the SRAM supports it), which would be excellent.
4. PSRAM tests are likely wrong. You need to evince the cache before starting the test, else you'll measure the time it takes to move from the cache to the cache. The default cache size in the ESP32S3 is 32KB (see [here](https://www.espressif.com/sites/default/files/documentation/esp32-s3_technical_reference_manual_en.pdf), p400 and later). In ESP-IDF, they usually configure it to 64KB. So the 100KB test already include a large part of it from the cache. You can use `esp_cache_msync` to evict it or simply read 64KB from somewhere here first to prune the cache, and then run the test. The size must also be larger than 128KB so when it finishes, the result is actually stored in the PSRAM and not in the cache. Else, what you'll see is like this:
```
Source PSRAM buffer
 32KB   |  32KB |  28KB  
[   A   |   B   |   C   ]

Dest PSRAM buffer  
 32KB   |  32KB |  28KB  
[   D   |   E   |   F   ]

Cache
[   x   |   x   ]

After warmup, you'll have this in the cache: (state diagram by time) 

|
|   S1    S2  S3    S4
+---*-----*----*----*--------------------> time

S1: Start of benchmark, cache contains [x|x]
S2: Read from PSRAM to cache and write to cache, cache contains [A|D] (actually measuring the PSRAM=>SRAM bandwidth for reading A + SRAM =>SRAM copying)
S3: End of reading A, starts reading B: doesn't fit. cache must evict its data to PSRAM, (actually measuring SRAM =>PSRAM bandwidth for transfering D to PSRAM, then PSRAM=>SRAM for reading B)
S4: End of reading C. Benchmarking stop, yet the data F isn't in PSRAM yet, since you haven't msync'ed yet.
```
The only part that would be relevant here for benchmark would be around state S3, since it's only there that the PSRAM is read and written. It happens after you've exhausted the cache and before it can leave lingering data in the cache, so after 32KB of source data and 32KB before its end. So in a 100KB test, you only experience this in the 28KB part between both ends. A 192KB test would have measured it in the inner 128KB area and left the 2 ends of 32KB unrelated.

Please notice that you can boost the PSRAM speed. On the ESP32-S3-WROOM-1U-N8R8, the PSRAM is mounted as a Octal SPI, and it's built by ESP-IDF by default as MSPI DDR mode running at 80MHz (see [here](https://docs.espressif.com/projects/esp-idf/en/stable/esp32s3/api-guides/flash_psram_config.html)). So the best bandwidth you could expect is 8bit * 80MHz * 2 = 160MB/s (a lot less, since there's the SPI and address selection overhead). Again, a copy operation will halve that (even more since you need to select the source and destination address for each transaction). So getting more than 50-60MB/s would be very excellent.

Yet, you can configure the MSPI to run at 120MHz (under the CONFIG_IDF_EXPERIMENTAL_FEATURES), with many caveats, but in that case you could expect a theorical bandwidth of 240MB/s or maybe ~100MB/s for a copy.

Anyway, thank you for your repository, it's excellent and I'm sure it can be even better. I've had hard time figuring out actual values for the DSP and SIMD stuff.

**Comment by project-x51** [comment]
> While reading the code, I'm spotting 3 potentials issues:

Thanks @X-Ryl669 

I modified the code to include the IRAM_ATTR prefix to any of the time sensitive code. This made a small improvement with the largest improvement (2.257%) being to the 'PIE 128-bit (32 byte loop) IRAM->IRAM' copy function. This now goes at 1,789.81 MB/s so not a bad way (its pretty much free) to go that little bit faster. After all every bit counts!

I'm still getting my head around the other two improvements you have suggested but wanted to post this so you knew I appreciated your comment.

I've created a Google Sheet to track improvements. You can find it [here](https://docs.google.com/spreadsheets/d/1A9UKdOb0QqLGQVSIru1gydPLEyCpcejhV0q_KI-OJMs/edit?usp=sharing). Please take a look if you have the time so I know its available to anyone interested.

**Comment by project-x51** [comment]
>Please notice that the memory bus of ESP32-S3 is 240MHz (32 bits, so a maximum theorical bandwidth is 960MB/s, a copy should halve that bandwidth so at best, you _should_ get 480MB/s). I don't understand how you could get more than that from the results you've shown. The ESP32 doesn't do any DDR like behavior on the SRAM, only the PSRAM supports it.
>    If your results are confirmed, then this means that the DSP/PIE can access the SRAM in larger format than 32 bits (and the SRAM supports it), which would be excellent.

@X-Ryl669 That would be the case if the ESP32 only had a 32bit internal memory bus. In fact its a 128bit data bus!! The only part of the processor that utilizes it is the PIE. Its capable of doing SIMD instructions such as 4 x 32 bit operations per clock cycle. Or as we use it a 1 x 128bit (16 bytes) load or store operation per clock cycle. This is how we manage to get the 1830.2 MB/s IRAM->IRAM result using the PIE. This is just shy of 4 x your theoretical max for moving 32 bits at 240 MHz.

**Comment by project-x51** [comment]
> 2. I'm not sure you're testing what you expect to test. The IRAM->IRAM test isn't doing what you think, since you've allocated the buffer with MALLOC_CAP_DMA and not MALLOC_CAP_32BIT. It's only with the latter that you _may_ get an IRAM buffer (see [here](https://docs.espressif.com/projects/esp-idf/en/stable/esp32s2/api-reference/system/mem_alloc.html#dma-capable-memory)). The other flag won't give you IRAM (they might give you DRAM, so you're measuring the SRAM speed here). 

@X-Ryl669 I took a look at the documentation you refer to on the **MALLOC_CAP_32BIT**. If you look just above it you will see that the DMA version excludes PSRAM but makes no mention of where else it might be. I don't see any mention of DRAM in the ESP32-S3 datasheet so I'm not sure how it could allocate DRAM, which only leaves IRAM. Anyway I tried changing the memory allocation to 32BIT and it made no difference. The results were identical for all the tests as best I can tell.

**Comment by project-x51** [comment]
> 3. PSRAM tests are likely wrong. You need to evince the cache before starting the test, else you'll measure the time it takes to move from the cache to the cache. The default cache size in the ESP32S3 is 32KB (see [here](https://www.espressif.com/sites/default/files/documentation/esp32-s3_technical_reference_manual_en.pdf), p400 and later). In ESP-IDF, they usually configure it to 64KB. So the 100KB test already include a large part of it from the cache. You can use `esp_cache_msync` to evict it or simply read 64KB from somewhere here first to prune the cache, and then run the test. The size must also be larger than 128KB so when it finishes, the result is actually stored in the PSRAM and not in the cache. Else, what you'll see is like this:

@X-Ryl669 Very interesting and not something I was aware of. I found more information by looking up the esp_cache_msync function [here](https://docs.espressif.com/projects/esp-idf/en/stable/esp32s3/api-reference/system/mm_sync.html). I'm not sure how to use this function yet as it wants me to specify memory starting addresses and size (no idea why) so I'll have a think on this one. I've also asked Microcontroller as he may have some more insight.

I did discover that the cache is only used when the CPU accesses PSRAM. It is bypassed for the DMA so some of the results may be skewed but the DMA ones should be correct. Defiantly a trap for someone not aware of this. I could just imagine trying to debug that one.

**Comment by project-x51** [comment]
> Yet, you can configure the MSPI to run at 120MHz (under the CONFIG_IDF_EXPERIMENTAL_FEATURES), with many caveats, but in that case you could expect a theorical bandwidth of 240MB/s or maybe ~100MB/s for a copy.

@X-Ryl669 I tried the 120 MHz PSRAM on my board. However it failed to get past the RAM detection code. It found it and posted information on the RAM and then hung. :(

I guess this is why its considered experimental. Note the warning says its not stable if the temperature moves by more than 20 degrees so I guess this makes it pretty useless except for some pretty niche use cases (if it worked).

Thanks for the suggestions. I'll let you know if I figure out how to properly test the PSRAM performance taking into account the cache load/flush challenges.

**Comment by X-Ryl669** [comment]
Thanks for your complete answer. As for the PSRAM performance measurement, I would do this way (pseudo code):
```cpp
char * inBuf = malloc_psram(192kB)
char * outBuf = malloc_psram(192kB)
// Pre fill the buffer
memset(inBuf, whatever)
// Start benchmarking 
clock false_start = now
copyBuffer(inBuf, outBuf, 32KB)
clock start = now
copyBuffer(inBuf+32KB, outBuf+32KB, 128KB)
clock end = now
copyBuffer(inBuffer+160kB, outBuf+160kB, 32kB)
clock false_end=now

// Mem benchmark speed:
totalWithCache = 192kB/(false_end - false_start)
totalWithoutCache = 128kB/(end - start)
totalPrePostLoad = 64kB/((false_end - end) + (start - false_start))
```

I've made the assumption that the cache is smart enough to split a 64kB R/W access in two independent 32kB access (one for read / one for write). Maybe I'm wrong, since there's no documentation about what algorithm is used on the caching process. Maybe the granularity is a lot smaller, so it'll be almost impossible to measure the effect of the caching, or it's larger (a single 64kB block?) in that case, the current test is already measuring the correct value. 

One hint however: if you memory map a flash page with MMU (which is, what I think the cache is using for PSRAM), the granularity is 64kB for ESP32S3, IIRC (see [here](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/system/mm.html) and [here for the actual page size](https://github.com/espressif/esp-idf/blob/master/components/soc/Kconfig))

**Comment by BitsForPeople** [comment]
>     1. Your code should be placed in IRAM (using the `IRAM_ATTR` [macro](https://docs.espressif.com/projects/esp-idf/en/stable/esp32s2/api-guides/memory-types.html#how-to-place-code-in-iram)) so it's not evicted from cache when the PSRAM is loading (and the time code doesn't measure slowness of SPI Flash reading).

The cache is split into code and data cache, so data cannot push out code, and vice versa. But yes, for maximum accuracy pre-loading the timed code into RAM would be best. Or running the same code more than once and taking the minimum run time.

>     2. I'm not sure you're testing what you expect to test. ... The other flag won't give you IRAM (they might give you DRAM, so you're measuring the SRAM speed here).

The -S3 actually has no DRAM (data RAM) or IRAM (instruction RAM); all its RAM is byte-addressable and can be executed from. (Except of course the RTC RAM, but we're not interested in that for now :))
So "IRAM" in this benchmark just means "internal RAM" a.k.a. SRAM, as opposed to the external PSRAM.
Adding MALLOC_CAP_INTERNAL to the allocations would make it extra clear that we want internal RAM.

>        If your results are confirmed, then this means that the DSP/PIE can access the SRAM in larger format than 32 bits (and the SRAM supports it), which would be excellent.

We confirmed that the results match what the TRM states, i.e. single-cycle 128-bit transfers from and to SRAM :)
(Actually, it's not 'fully' single-cycle: While a 128-bit load takes the CPU 1 clock cycle to execute, the data loaded is only available in the respective CPU register 1 clock cycle later, so worst case is 2 CPU clock cycles to load 128 bits; but by exploiting the CPU's pipeline you can achieve a throughput of 128 bits per clock cycle. That's why loading 2x16 bytes, then storing 2x16 bytes is faster than 2x loading+storing 16 bytes)

**Comment by BitsForPeople** [comment]
> I've made the assumption that the cache is smart enough to split a 64kB R/W access in two independent 32kB access (one for read / one for write). Maybe I'm wrong, since there's no documentation about what algorithm is used on the caching process. Maybe the granularity is a lot smaller, ...

It is a lot smaller. The granularity is one cache line, which can be configured to be 32 or 64 bytes.

**Comment by BitsForPeople** [comment]
To avoid the effects of the cache when measuring non-DMA transfers from/to PSRAM, we would have to 

1. esp_cache_msync( sourceBuffer, sourceSize, ESP_CACHE_MSYNC_FLAG_DIR_C2M ); // Make sure the source data we want to copy has actually made it to PSRAM
2. esp_cache_msync( sourceBuffer, sourceSize, ESP_CACHE_MSYNC_FLAG_DIR_M2C ); // Make the cache forget the cached source data
3. esp_cache_msync( destBuffer, destSize, ESP_CACHE_MSYNC_FLAG_DIR_M2C ); // Make the cache forget any data in the destination it may still want to write to PSRAM
4. start timing
5. copy data from/to PSRAM
6. esp_cache_msync( destBuffer, destSize, ESP_CACHE_MSYNC_FLAG_DIR_C2M ); // Make sure the data we copied to dest has actually made it to PSRAM
7. stop timing

Here, 1. and 2. only apply when we _read from_ PSRAM; 6. only applies when we _write to_ PSRAM. 3. applies if there may be some dirty data in the cache for destBuffer; it's redundant if we're sure that's not the case, e.g. because 6. was previously executed.

Edit:
It seems that 1. and 2. can be combined into a single esp_cache_msync( sourceBuffer, sourceSize, ESP_CACHE_MSYNC_FLAG_DIR_C2M | ESP_CACHE_MSYNC_FLAG_INVALIDATE ).

To be extra accurate, we should actually evict _all_ PSRAM data from the cache first. Otherwise there may be some dirty PSRAM data from a previous run in the cache, which will likely get pushed out of the cache during our test, forcing the cache to write it back to PSRAM while we're timing the test.

**Comment by X-Ryl669** [comment]
> It is a lot smaller. The granularity is one cache line, which can be configured to be 32 or 64 bytes.

I think we are not speaking about the same thing here. The cache line is 64 bytes on ESP32-S3. But this doesn't matter for the test. What matter is the cache block size (in SRAM) which is 32K or 64K since when you're benchmarking the memory copy, you're are actually measuring the time it takes to copy from SRAM to SRAM if the data you're copying to/from can fit in this cache block size. You need to overcome the block size to actually measure the PSRAM speed. 

When using the CPU (not DMA), you can't skip the caching mechanism, so you must measure the time it takes to fill / empty the cache too, which the current code didn't do (the filling of the cache was here obviously, but the emptying wasn't, since the test stopped without knowing when the cache lines were transmitted). This is relevant to a real world usage. I don't know if the caching algorithm was dumb (like, every time a cache line is touched it'll be marked as dirty and will be transferred later on, and in that case, the benchmark is almost correct, except for the last cache line) or a bit more smarter (like, until all the cache block is used, no cache line is transferred to PSRAM and then only when new data are loaded in the cache that'll trigger transferring the stale/old data to PSRAM, and in that case, the benchmark is irrelevant, since it was fitting in the cache block size anyway).

From the post of Sprite_TM [here](https://esp32.com/viewtopic.php?t=24459&start=90#p94820) it seems it's the latter, so it's definitively an issue in the benchmark.

However, since the DMA bypass the cache, the actual PSRAM speed is captured here (althrough it also capture a bit of the FreeRTOS's async scheduling overhead so for such a small buffer, it matters too). 

I think to get a better idea of the memcpy speed from & to PSRAM, a larger block is required in all cases (cache or no cache) so that the caching overhead (**which is useless in this usage pattern since data is never used twice**) or async scheduling overhead is minimized.

**Comment by X-Ryl669** [comment]
> To avoid the effects of the cache when measuring non-DMA transfers from/to PSRAM, we would have to [...]

Yes that's correct.

**Comment by BitsForPeople** [comment]
With proper(?) cache flushing the results look different w.r.t. IRAM<->PSRAM speed:

`I (931) Memory Copy: 8-bit for loop copy IRAM->PSRAM took 796811 CPU cycles = 29.41 MB/s`
`I (981) Memory Copy: 16-bit for loop copy IRAM->PSRAM took 460818 CPU cycles = 50.86 MB/s`
`I (1031) Memory Copy: 32-bit for loop copy IRAM->PSRAM took 403215 CPU cycles = 58.13 MB/s`
`I (1081) Memory Copy: 64-bit for loop copy IRAM->PSRAM took 412809 CPU cycles = 56.78 MB/s`
`I (1131) Memory Copy: memcpy IRAM->PSRAM took 412836 CPU cycles = 56.77 MB/s`
`I (1181) Memory Copy: async_memcpy IRAM->PSRAM took 461617 CPU cycles = 50.77 MB/s`
`I (1231) Memory Copy: PIE 128-bit (16 byte loop) IRAM->PSRAM took 403207 CPU cycles = 58.13 MB/s`
`I (1281) Memory Copy: PIE 128-bit (32 byte loop) IRAM->PSRAM took 403209 CPU cycles = 58.13 MB/s`
`I (1331) Memory Copy: DSP AES3 IRAM->PSRAM took 405053 CPU cycles = 57.86 MB/s`

`I (1381) Memory Copy: Swapping source and destination buffers`
`I (1431) Memory Copy: 8-bit for loop copy PSRAM->IRAM took 864096 CPU cycles = 27.12 MB/s`
`I (1481) Memory Copy: 16-bit for loop copy PSRAM->IRAM took 736145 CPU cycles = 31.84 MB/s`
`I (1531) Memory Copy: 32-bit for loop copy PSRAM->IRAM took 717695 CPU cycles = 32.66 MB/s`
`I (1581) Memory Copy: 64-bit for loop copy PSRAM->IRAM took 717694 CPU cycles = 32.66 MB/s`
`I (1631) Memory Copy: memcpy PSRAM->IRAM took 720810 CPU cycles = 32.52 MB/s`
`I (1681) Memory Copy: async_memcpy PSRAM->IRAM took 444946 CPU cycles = 52.67 MB/s`
`I (1731) Memory Copy: PIE 128-bit (16 byte loop) PSRAM->IRAM took 720756 CPU cycles = 32.52 MB/s`
`I (1781) Memory Copy: PIE 128-bit (32 byte loop) PSRAM->IRAM took 720758 CPU cycles = 32.52 MB/s`
`I (1831) Memory Copy: DSP AES3 PSRAM->IRAM took 722728 CPU cycles = 32.43 MB/s`

**Comment by X-Ryl669** [comment]
I suspect the "cache flushing" function to simply trash the cache lines by filling them with random memory addresses. I need to do my homework here and read them.

I don't get why you are flushing the memory for async_memcpy since DMA doesn't use the cache anyway. 

The value are consistent with the architecture here, PSRAM is using 8 bits/cycle at a 80MHz clock in DDR fashion, so at best it can do 160MB/s (thus at best 80MB/s for a memory copy to/from PSRAM without cache).  Since you need to select the address for transferring from the chip (10 cycles per transfer, provided it's a burst one with 64 bytes, so 32 cycles in DDR, that's a 1/3 overhead here), it's a lot less than that, so a maximum theoretical value of ~60MB/s is achievable (80MB/s * 32 cycles/42 cycles). Reaching a value of ~50MB/s doesn't look too far from the reality.  

It's good to see that the effect of caching actually improve performance (38MB/s instead of 32MB/s), since this proves that the address selection overhead in the SPI memory is important. It's not as important as expected (I would have expected a 1/3 increase here) but it's nonetheless useful.

**Comment by BitsForPeople** [comment]
I'm only wondering why the DMA PSRAM->IRAM transfer would be so much faster than the CPU-driven one.
Seeing that, I added cache prepping around the DMA too, but as expected it made no difference.

I assume the DMA, knowing the size and sequential nature of the transfer beforehand, uses longer bursts than the cache, reducing the relative overhead.

**Comment by BitsForPeople** [comment]
> I suspect the "cache flushing" function to simply trash the cache lines by filling them with random memory addresses.

No. The newer ESPs' caches support selectively flushing, invalidating, prefetching, and locking address ranges.

**Comment by project-x51** [comment]
I've merged the code changes and added a flag to compile with the cache code or without it. I get lower results for some of the PSRAM->IRAM copies (to be expected I guess) and no real change in the IRAM->PSRAM (Strange).

Also strange I get significantly faster results for the 8 bit loop copies. Especially for the IRAM->IRAM which makes absolutely no sense. Any ideas?

**Comment by project-x51** [comment]
<img width="903" alt="{5E5E28E8-E69F-4D74-B670-71CC7435A99E}" src="https://github.com/user-attachments/assets/fc1f2659-c24f-4365-b76f-b4c04c3f67c1">

**Comment by X-Ryl669** [comment]
I think the title "With cache" and "No cache" are inverted in the column, IIRC. The new code *tries* to disable the effect of the cache, not enabling it.

And for the 8 bit loop copy speedup, I have no conclusions either. You might want to compare the disassembly of the both code to see what has changed. Technically, there's nothing that prevent the compiler to write more efficient code (like ignoring you're doing 8bit->8bit moves/copy and instead write it as a 32bit->32bits move or more). Also, it's possible it is unrolling the loop, or the alignment for the pointers in the new version are now a multiple of 16 bytes allowing more efficient code, or the compiler version increased, or the memory barrier introduced in the new code allowed the compiler to take a shortcut that it couldn't do before (it would have been unsafe doing so before and now it's ok).

**Comment by BitsForPeople** [comment]
Another PR which changes some results: #4 :)

**Comment by project-x51** [comment]
> I think the title "With cache" and "No cache" are inverted in the column, IIRC. The new code _tries_ to disable the effect of the cache, not enabling it.

I've sorted this now. I'm using the terms "With Cache Flush" and "No Cache Flush"

**Comment by project-x51** [comment]
> Another PR which changes some results: #4 :)

Sorted. I've changed the spreadsheet so the results still make sense historically. The biggest change is to the async_memcpy IRAM->PSRAM. No change in the other direction. 

These results make sense now. Going to PSRAM the flush will matter. In the other direction the device has all the time in the world to move the data into IRAM as it arrives and I doubt it even uses the cache.

![image](https://github.com/user-attachments/assets/c5074376-4fe7-423b-82ba-9034ae7fcf0d)

**Comment by X-Ryl669** [comment]
Makes a lot more sense now.

**Comment by project-x51** [comment]
Thanks @X-Ryl669 and @BitsForPeople. I'll close this issue now.

**Comment by Decryptortuning** [comment]
Have you considered attempting to measure latency? Could be a cool addition to the benchmark.
There have been some huge improvements in JTAG sample rates that may be of interest.
https://esp32.com/viewtopic.php?f=13&t=39619

I am excited to dig into p4 optimizations soon.

**Comment by BitsForPeople** [comment]
Not sure exactly what latency you mean and how JTAG (profiling?) would play a role, but:
On the S3, reads from internal RAM have a latency of 1 CPU clock cycle. The cache is in internal RAM so same here.
In comparison, data pulled from external RAM will have much higher latency; that might be interesting to measure.

**Comment by project-x51** [comment]
We could use esp_timer_get_time() to measure latency to the microsecond.

**Comment by BitsForPeople** [comment]
As to the PSRAM latency, we can also calculate it from the data we have:
PSRAM at 80MHz OPI transfers 80MB/s from PSRAM to the ESP, but we measure only ~58MB/s; the difference can be defined as the latency.
Cache line ("block") size is 32 bytes, so one cache line of data takes 400ns for the actual data transfer (32 bytes @ 80MB/s).
The 58MB/s we measured means that loading a 32-byte block actually took ~550ns, i.e. 400ns for the transfer + 150ns of 'latency'.
As the cache transfers full blocks/lines only, reading one byte from PSRAM incurs the same latency as reading one line (of 32 bytes in this configuration) from PSRAM (except for the tiny amount of additional latency for reading from the cache).

**Comment by X-Ryl669** [comment]
Don't forget the address selection overhead. On PSRAM you must select the address first and then the data to read or write. This consumes ~1/3rd of the bandwidth, so the 58MB/s is already quite good.

**Comment by BitsForPeople** [comment]
Yes, that and all other overhead is included in the 150ns. I expect the 150ns to be the fixed 'cost', independent of the cache line size. Thus, with 16-byte cache lines we should see 150+200ns until data is available from PSRAM (~45MB/s max.), and 150+800ns with 64-byte lines (~67MB/s max.). This total time can also be considered as the 'latency', i.e. the time from a read being issued by the CPU until the first data arrives at the CPU.

**Comment by BitsForPeople** [comment]
Yeah - no. It's not as simple as I had hoped.
Here's what I get:

Cache line size **16**:
`I (1881) Memory Copy: PIE 128-bit (32 byte loop) PSRAM->IRAM took 633664 CPU cycles = 36.99 MB/s`

Cache line size **32**:
`I (1791) Memory Copy: PIE 128-bit (32 byte loop) PSRAM->IRAM took 403334 CPU cycles = 58.11 MB/s`

Cache line size **64**:
`I (1791) Memory Copy: PIE 128-bit (32 byte loop) PSRAM->IRAM took 278493 CPU cycles = 84.16 MB/s`

So much for the "80MB/s" theoretical limit...

Datasheet says "Up to 8-line SDR/DDR (Single Data Rate/**Double Data Rate**) reads and writes are supported".
Anyone knows how I can figure out if my chip uses SDR or DDR?

**Comment by BitsForPeople** [comment]
The three datapoints are consistent with a fixed latency of ~327.7ns and a transfer rate of 147MB/s (~6.8ns/byte). So DDR it must be?

**Comment by X-Ryl669** [comment]
Yes, there isn't any SDR mode on ESP32 for PSRAM, they are [all DDR](https://github.com/espressif/esp-idf/issues/10417#issuecomment-1366363401). However, in some [PSRAM protocol](https://www.infineon.com/dgdl/Infineon-S27KL0643_S27KS0643_3.0_V_1.8_V_64_MB_(8_MB)_HyperRAM_Self-Refresh_DRAM-DataSheet-v06_00-EN.pdf?fileId=8ac78c8c7d0d8da4017d0ee8b87c7185), sending the address requires 6 clock cycles, plus the actual latency to select the column / row of the memory itself (min 4 cycles, can be 8 cycles if the chip requires 2x latency). So the total latency before actual data is at minimum 10 cycles (125ns at 80MHz or 175ns if double latency). Then there's the data itself, 400ns for 64 bytes (32 cycles). So a grand minimum total of 525ns to 575ns for 64 bytes => 121MB/s to 111MB/s.

Getting 84MB/s instead of 111MB/s gives a 3/4 of actual bandwidth. Which is quite good actually.

**Comment by BitsForPeople** [comment]
![Image](https://github.com/user-attachments/assets/2fac2de5-c84d-4d11-bb5c-f2c1281ad512)

If we subtract the 0.125 CPU cycles (0.52ns) per byte (2 cycles for 16 bytes) needed to read from the cache and write to IRAM, we end up with 159.5MB/s for the transfer speed PSRAM->cache. Now if that's not close to 80MHz DDR... ;-) Nice.

**Comment by X-Ryl669** [comment]
> we end up with 159.5MB/s for the transfer speed PSRAM->cache

That doesn't make sense to me. It's physically impossible, you need to select the address first and doesn't match the expected serial RAM protocol. You can only get 160MB/s on a 80MHz DDR SPI if the chip was streaming data non stop. But it doesn't do that, there are pauses because the SPI RAM isn't a true SRAM, it takes time to select the address to fetch. The observed data match this fact, you have incompressible latency when accessing the PSRAM and then you have a burst of data that's running at 160MB/s (and only that part), since it can't be any other way, the chip **must** give a byte on each clock edge on a DDR octal data bus.

**Comment by BitsForPeople** [comment]
The 159.5MB/s is only the calculated data transfer rate _after_ the (fixed) 328ns latency period; all address setup and other fixed overhead is included in the 328ns. (And with the latency included, we only get the 84MB/s overall the benchmark displays.)
I did that calculation to check if the 328ns is actually correct: If 328ns is correct, the remaining time of the whole operation should correspond to 160MB/s at 80MHz DDR; and it does when we remove the known CPU 'overhead', so I'm confident now that we have pretty accurate numbers figured out.

**Comment by BitsForPeople** [comment]
And this is what you get when you remember that the benchmark prints its results in (1024*1024)-byte megabytes, not 1000000... ;-)

![Image](https://github.com/user-attachments/assets/008e9e94-7081-4973-b60d-07023e06205c)
